bop_toolkit_lib/score.py

```diff
@@ -177,158 +177,135 @@ def calc_pose_detection_scores(
-                    detection_types["false_negative_ignore"][i] = True
+                    ignored_true_positive_mask[match_id] = True
             else:
-                # Case 3 (false_positive): m["est_id"] == -1 and not m["valid"]
-                detection_types["false_positive"][i] = True
+                if match["gt_visib_fract"] >= visib_gt_min:
+                    false_negative_mask[match_id] = True
+                else:
+                    ignored_false_negative_mask[match_id] = True
     
-        if double_check_size:
-            # Double check whether the size of GT is correct
-            num_gts = len([m for m in obj_matches if m["valid"]])
-            assert num_gts == np.sum(detection_types["true_positive"]) + np.sum(detection_types["false_positive_ignore"]) + np.sum(detection_types["false_negative_ignore"]) + np.sum(detection_types["false_negative"])
-
-        # remove the detections that should be ignored
-        ignored_idex = np.logical_or(detection_types["false_negative_ignore"], detection_types["false_positive_ignore"])
-        keep_idx = np.invert(ignored_idex)
-        true_positives = detection_types["true_positive"][keep_idx]
-        false_positives = detection_types["false_positive"][keep_idx]
-
-        # Recall
-        # Update the number of GT for reccall
-        obj_tar = obj_tar - np.sum(detection_types["false_negative_ignore"])
-        recall = cum_true_positives / int(obj_tar)
-
-        # Precision
-        cum_true_positives = np.cumsum(true_positives)
-        cum_false_positives = np.cumsum(false_positives)
-        precision = cum_true_positives / (cum_true_positives + cum_false_positives)
-        precision[np.isnan(precision)] = 0
-
-        ap = calc_ap(recall, precision, coco_interpolation=True)
-        scores_per_object[obj_id] = ap
-        num_instances_per_object[obj_id] = int(obj_tar)
+        # Make sure each GT was labeled.
+        assert num_obj_gts == (
+            np.sum(true_positive_mask) +
+            np.sum(ignored_true_positive_mask) +
+            np.sum(false_negative_mask) +
+            np.sum(ignored_false_negative_mask)
+        )
+
+        # Consider only valid true positive and false negative detections (valid = not associated
+        # with any ignored GT).
+        ignored_mask = np.logical_or(ignored_true_positive_mask, ignored_false_negative_mask)
+        kept_mask = np.invert(ignored_mask)
+        true_positive_mask = true_positive_mask[kept_mask]
+        false_negative_mask = false_negative_mask[kept_mask]
+
+        # Number of valid detections for the current object.
+        obj_dets_count = len([err for err in errs if err["obj_id"] == obj_id])
+
+        # Number of valid detections (valid = not associated with any ignored GT).
+        valid_obj_dets_count = obj_dets_count - np.sum(ignored_true_positive_mask)
+        assert valid_obj_dets_count >= 0
```

Comment:
It might happen that the submissions do not have any valid detections for some object IDs, should we remove this condition ?

Reply:
Then `valid_obj_dets_count` should be 0 and thus the assert should be satisfied. This assert is there to check whether `valid_obj_dets_count` is not negative (which could happen if something went wrong).